### PR TITLE
Fixes and adoptions, custom pre/post actions

### PR DESCRIPTION
**Pull request by mikhailnov**
Hi, thanks for this script, I have adopted it a bit for my needs. What I have done is not finished, but it works for my needs and may be will intreset you.

* added ability to set custom `btrfs` command
* added ability to set custom `btrfs` commands in send/receive pipeline, I thought to use it for SSH but then undestood that rewriting the logics of how existence of snapshots is checked is required, so it exists but may be useless
* prefix `history.db` filename with a custom name: this fixes the work when there are multiple snapshottable subvolumes in one root
* added pre/post actions to run before and after the snapshot.

I use pre/post snapshot actions to mount and umount a root subvolume inside which there are multiple snapshottable subvolumes. I tried to add a trap to catch any exit of the script, including the ones when it exists somewhere in the middle dut to errors to guarantee unmounting what was mounted as a pre-snapshot action, but faced a problem that it does not work well when multiple snapshots are defined in one json config. And I faced some other problems when multiple snapshots are defined in one json config, already do not remember details.
As I am not good in php, I ended up with a shell script like `for i in config1.json config2.json config3.json; do just-backup-btrfs "$i" ; done`.

Example of my json config:
```
[
	{
		"snapshot_name_prefix"	       : "gitea",
		"action_snapshot_pre"          : "mkdir -p /tmp/_jbb/gitea/ && chmod 700 /tmp/_jbb/gitea/ && mount /dev/disk/by-uuid/xxxxxxx /tmp/_jbb/gitea/",
		"action_snapshot_post"         : "umount /tmp/_jbb/gitea/",
		"source_mounted_volume"        : "/tmp/_jbb/gitea/@gitea",
		"destination_within_partition" : "/tmp/_jbb/gitea/jbb_snaps/gitea/",
		"destination_other_partition"  : null,
		"date_format"                  : "Y-m-d_H:i:s",
		"btrfs_cmd"		               : "btrfs",
		"skip_auto_mounting"           : "true",
		"keep_snapshots"               : {
			"hour"  : 6,
			"day"   : 24,
			"month" : 30,
			"year"  : 48
		}
	}
] 
```

Plus I use `btrfs-sync` script https://github.com/nachoparker/btrfs-sync to send incremental backups to another server.

### CODE REVIEW

**Comment by nazar-pc** [comment]
First of all, this is a lot of things for a single PR.
As for the features:
* I don't see the point of having a custom `btrfs` command
* I don't really understand why would you ever want to have snapshots of different subvolumes in the same directory, it makes no sense to me (just create appropriate number of subdirectories/subvolumes)
* again, do not really understand why you need pre/post hooks for mounting something (I'd expect it to be mounted all the time or mounted before calling just-backup-btrfs), but this seems to be a potentially useful feature nevertheless

If you can decouple pre/post hooks separately such that we can work on them separately and merge eventually, that would be nice.

As for SSH, btrfs wiki has links to several solutions, maybe some will suit your needs. I didn't need SSH and wanted something simple and manageable, so I wrote this and use it successfully for many years now.

**Comment by mikhailnov** [comment]
> If you can decouple pre/post hooks separately such that we can work on them separately and merge eventually, that would be nice.

I think it will be impossible to split changes into different PRs because they will conflict with each other and be umergeable.

> * I don't see the point of having a custom `btrfs` command

most of all agree with you, I made it because I made custom commands for send|receive pipelines

> * I don't really understand why would you ever want to have snapshots of different subvolumes in the same directory, it makes no sense to me (just create appropriate number of subdirectories/subvolumes)

That is the layout that I use everywhere, an example:
/dev/sda1 is btrfs
system is installed into subvolume `@`
/home is in subvolume `@home`
(that is what Ubiquity, Ubuntu-Debian console and Calamares installers do by default)
The root subvolume of /dev/sda1 is not mounted, only subvolumes `@` and `@home` are mounted to `/` and /`home`.
I use the same layout on server for systemd-nspawn containers. From `/etc/fstab` as an example:
```
UUID=xxx /var/lib/machines/webserver1 btrfs subvol=@webserver1_vm,compress=zstd
UUID=xxx /var/lib/machines/bbb1 btrfs subvol=@bbb1,compress=zstd
UUID=xxx /var/lib/machines/gitea btrfs subvol=@gitea,compress=zstd
```
`UID=xxx` is the same in all those lines.
I can do:
```
mkdir -p /tmp/0
mount /dev/sda1 /tmp/0
btrfs subvol snapshot -r /tmp/0/@gitea /tmp/0/@gitea_snapshot
umount /tmp/0
```

This allows to have the root of the subvolume and snapshots inaccessible for additional security.
I also find using the root of the subvolume directly a bad idea because it requires very tricky to understand layouts of snapshots.

Snapshots are not accessible without mounting. That is an improtant idea.

I found making custom mount commands much easier, better understandable and more predictable than trying to adopt automounting logics of this scripts for my disk layout.

**Comment by nazar-pc** [comment]
> I think it will be impossible to split changes into different PRs because they will conflict with each other and be umergeable.

Of course it is possible.

> system is installed into subvolume @
> /home is in subvolume @home

This tool already supports such layout. Just use `/` and `/home` as `source_mounted_volume` and the tool will figure out the correct filesystem root and create snapshot in location you desire.

> This allows to have the root of the subvolume and snapshots inaccessible for additional security.

This is not a security, this is some kind of obscurity. Moreover, snapshots are read-only, so someone will need root access to delete them. At which point you have much bigger issues than accessible snapshots.

Also do you really need hooks, can't you just do?:
```bash
/whatever-before-making-snapshot
just-backup-btrfs
/whatever-after-making-snapshot
```
In other words, doing one command before all snapshots and one command after all snapshots rather than doing it on each snapshot.

**Comment by nazar-pc** [comment]
If customized script works for you that is great, but I wouldn't merge it as is.
Hence closing PR, but feel free to continue discussion.
